### PR TITLE
New sdba module for Bias-adjustment (and statistical downscaling)

### LINKED ISSUE #423: Basic downscaling methods
### Description

Implement basic downscaling methods (1D) typically used in the production of scenarios.

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #423
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [x] `bumpversion (minor / major / patch)` has been called on this branch
- [ ] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** 

Title says it all. A whole new xclim module, with bias-adjustment methods and all the tools to adjust climate data.

* **Does this PR introduce a breaking change?** 

No, but the module requires a _future_ version of xarray : the polynomial detrending uses methods not yet published. Using xarray at master works and importing `xclim.sdba` will fail with the current xarray version available in the repos.

* **Other information**:

@huard @tlogan2000
    - **Calendars** : I "fixed" the dayofyear dilemma by interpolating to the largest doy range in `get_correction`, also raising a warning.  Once my other PR is merged (code already there but commented), the plan is to do the same directly in the `train` and `adjust` calls if the input calendar and the one of `sim` do not fit (only for "time.dayofyear" grouping).
    - **Naming** : I left `apply_correction` and `get_correction`. I felt that `adjustment` was too powerful to name a simple addition/subtraction, correction is more general.

### CODE REVIEW

**Comment by huard** [comment]
It could be months before the next xarray release. I suggest we find a way to merge this without causing import failures with the current release.

File: xclim/sdba/adjustment.py [hunk 1]
```diff
@@ -0,0 +1,349 @@
...
+
+class EmpiricalQuantileMapping(BaseAdjustment):
+    """Empirical Quantile Mapping bias-adjustment.
+
+    Adjustment factors are computed between the quantiles of `ref` and `sim`.
+    Values of `sim` are matched to the corresponding quantiles of `hist` and corrected accordingly.
+
+    .. math::
+
+      F^{-1}_{ref} (F_{mod}(x_{mod}))
```
Comment (huard) [inline xclim/sdba/adjustment.py:109]:
```suggestion
      F^{-1}_{ref} (F_{hist}(sim))
```

File: xclim/sdba/adjustment.py [hunk 2]
```diff
@@ -0,0 +1,349 @@
...
+
+class DetrendedQuantileMapping(EmpiricalQuantileMapping):
+    """Detrended Quantile Mapping bias-adjustment.
+
+    A scaling factor that would make the mean of `hist` match the mean of `ref` is computed.
+    `ref` and `hist` are normalized by removing the group-wise mean.
+    Adjustment factors are computed between the quantiles of the normalized `ref` and `hist`.
+    `sim` is corrected by the scaling factor then detrended using a linear fit.
+    Values of detrended `sim` are matched to the corresponding quantiles of normalized `hist` and corrected accordingly.
+    The trend is put back on the result.
```
Comment (huard) [inline xclim/sdba/adjustment.py:188]: Either convert to complete paragraph, or to a numbered list.

File: xclim/sdba/adjustment.py [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+
```
Comment (huard) [inline xclim/sdba/adjustment.py:189]: Add equation from paper.

File: xclim/sdba/adjustment.py [hunk 4]
```diff
@@ -0,0 +1,349 @@
...
+
+class QuantileDeltaMapping(EmpiricalQuantileMapping):
+    """Quantile Delta Mapping bias-adjustment.
+
+    Adjustment factors are computed between the quantiles of `ref` and `hist`.
+    Quantiles of `sim` are matched to the corresponding quantiles of `hist` and corrected accordingly.
+
+    The algorithm is based on the "QDM" method of [Cannon2015]_.
+
```
Comment (huard) [inline xclim/sdba/adjustment.py:230]: Add equation from paper.

File: xclim/sdba/adjustment.py [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+        af = broadcast(af, sim, group=self.group, interp=self.interp, sel=sel)
+
+        return apply_correction(sim, af, self.kind)
+
+
+class LOCI(BaseAdjustment):
+    r"""Local Intensity Scaling (LOCI) bias-adjustment.
+
+    This bias adjustment method is designed to correct daily precipitation time series by considering wet and dry days
+    separately. Based on [Schmidli2006]_.
```
Comment (huard) [inline xclim/sdba/adjustment.py:250]:
```suggestion
    separately ([Schmidli2006]_).
```

File: xclim/sdba/adjustment.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+    The threshold on the training target `ref` is first mapped to `hist` by finding the quantile in `hist` having the same
+    exceedance probability as thresh in `ref`. The adjustment factor is then given by
+
+    .. math::
+
+       s = \frac{\left \langle ref: ref \geq t_{ref} \right\rangle - t_{ref}}{\left \langle hist : hist \geq t_{hist} \right\rangle - t_{hist}}
+
+    In the case of precipitations, the adjustment factor is the ratio of wet-days intensity.
+
+    For a adjustment factor `s`, the bias-adjustment of `sim` is:
```
Comment (huard) [inline xclim/sdba/adjustment.py:264]:
```suggestion
    For an adjustment factor `s`, the bias-adjustment of `sim` is:
```

File: xclim/sdba/base.py [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+            prop=prop,
+            name=group,
+            window=window,
+            interp=interp,
+        )
+
+    def group(self, da: xr.DataArray = None, **das: xr.DataArray):
+        """Return a xr.core.groupby.GroupBy object.
+
+        More than one arrays can be combined to a dataset before grouping using the `das`  kwargs.
```
Comment (huard) [inline xclim/sdba/base.py:108]:
```suggestion
        More than one array can be combined to a dataset before grouping using the `das`  kwargs.
```

**Comment by aulemahal** [comment]
This last commit solves the problem with xarray not being at master in the main test builds. However, our only test suite testing sdba is not the @master one, which still fails because of a bug in xArray.

**Comment by huard** [comment]
The tests in sdba are not picked up by pytest. Rename the directory to test_sdba ?

**Comment by aulemahal** [comment]
They are picked up! But I added `importorskip` statements, since `sdba` cannot be imported with the current version of xarray. If you check the "@master" travis build, the sdba tests pass. But there are other errors related to a xarray bug.

**Comment by huard** [comment]
Argh ! Should have thought about this. Thanks!

**Pull request merged**